### ISSUE TITLE
[labs/react] Typescript errors with react wrapped component in Vite Preact Typescript template

### ISSUE THREAD

**Issue by ajmchambers** [lit/lit#4138]
### Which package(s) are affected?

React (@lit-labs/react)

### Description

Trying out shoelace in a fresh vite preact-ts template and getting a couple of typescript issues.

(shoelace 2.8.0 which is using @lit-labs/react 2.0.1, NOTE: react wrapper files are generated on build [here](https://github.com/shoelace-style/shoelace/blob/next/scripts/make-react.js#L58-L68))

With the following code I am getting 2 different typescript errors:

```tsx
import { useState } from 'preact/hooks';
import SlButton from '@shoelace-style/shoelace/dist/react/button/index.js';
import SlIcon from '@shoelace-style/shoelace/dist/react/icon/index.js';

export default function Test() {
    const [count, setCount] = useState(0);
    return (
        <SlButton onClick={() => setCount((count) => count + 1)} variant="primary">
            <SlIcon slot="prefix" library="system" name="star-fill" />
            count is {count}
        </SlButton>
    );
}
```

SlButton type error (not an error in stackblitz example but it is locally in vscode):
```
'SlButton' cannot be used as a JSX component.
  Its type 'ReactWebComponent<SlButton, { onSlBlur: EventName<SlBlurEvent>; onSlFocus: EventName<SlFocusEvent>; onSlInvalid: EventName<SlInvalidEvent>; }>' is not a valid JSX element type.
    Type 'ForwardRefExoticComponent<Omit<HTMLAttributes<SlButton>, "onSlBlur" | "onSlFocus" | "onSlInvalid"> & Partial<Omit<SlButton, keyof HTMLElement>> & EventListeners<...> & RefAttributes<...>>' is not assignable to type 'FunctionComponent<any>'.
      Type 'ReactNode' is not assignable to type 'VNode<any> | null'.
        Type 'undefined' is not assignable to type 'VNode<any> | null'.ts(2786)
```

SlIcon has this error:
```
Type 'Element' is not assignable to type 'ReactNode'.
  Property 'children' is missing in type 'VNode<any>' but required in type 'ReactPortal'.ts(2322)
```

If I change SlIcon to a div:
```tsx
<div slot="prefix"></div>
```

The typescript error is just:
```
Type 'Element' is not assignable to type 'ReactNode'.ts(2322)
```

### Reproduction

Here is a stackblitz example (note: the first error does not appear here for me, only in local vscode):
https://stackblitz.com/edit/vitejs-vite-k7w2nf?file=src%2Fapp.tsx&terminal=dev

### Workaround

I have not found a workaround

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

2.0.1

### Browser/OS/Node environment

Browser: Edge 116.0.1938.54
OS: Windows 11 22621.2215
Node version: 18.11.0
Npm version: 9.2.0

**Comment by augustjk** [comment]
The problem arises because our wrapped components are typed to have their `children` typed as `React.ReactNode`. This works fine in React projects, but in Preact projects any elements or components you place in children position, even if it's another React component, are typed differently and incompatible.

The solution would be to explicitly type our component's `children` prop to accept renderable child values. This would be `ReactNode` for React, and `ComponentChild` for Preact. Using `JSX.Element` would cover the compatibility between `ReactElement` & `VNode`, and we can add primitives that React accepts, string, number, boolean, null, undefined. Preact also accepts object and bigint, but we won't accept those.

**Comment by augustjk** [comment]
@ajmchambers Actually.. I think I found a better solution that doesn't rely on the `@lit-labs/react` type to do anything special.

As documented here https://preactjs.com/guide/v10/typescript/#typescript-preactcompat-configuration, relying the `preact/compat` types would alleviate all this.

Adding just the `"paths"` entry to the stackblitz repro you provided:
https://stackblitz.com/edit/vitejs-vite-yhasfk?file=tsconfig.json,src%2Fapp.tsx

**Comment by ajmchambers** [comment]
Hmm not sure its that simple unfortunately.

The vite preact preset has preact/compat aliases defined by default at least it mentions it here: https://github.com/preactjs/preset-vite

I tested your stackblitz and there are no errors, but there is also no auto completion. For example try setting the variant attribute on SlButton. Or setting props that don't exist like fish="carp" gives no errors.

**Comment by augustjk** [comment]
I think the vite preact preset only adds the compat in the babel config which isn't involved in type checking. Adding to tsconfig does add it for type checking but it seems `preact/compat` doesn't include a type for `PropsWithoutRef`, making the prop basically `any`.

Until `preact` adds the type, we could perhaps ship our own copy of it.

**Comment by AndrewJakubowicz** [comment]
The copy of the type has been released in [@lit-labs/react@2.0.3](https://github.com/lit/lit/releases/tag/%40lit-labs%2Freact%402.0.3)
